### PR TITLE
tests(storage): add universe domain integration test

### PR DESCRIPTION
**Pull request by ddelgrosso1**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14728)

### CODE REVIEW

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14728)* status: 0 of 7 files reviewed, 1 unresolved discussion (waiting on @ddelgrosso1)

___
*[`ci/cloudbuild/builds/universe-domain-storage.sh` line 29 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14728#-O7_kQHH1Dc8cStlfUc-:-O7_kQHH1Dc8cStlfUc0:bts0b6l) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/13d08a440771d288e8f24db49f1c56ac34c06f8b/ci/cloudbuild/builds/universe-domain-storage.sh#L29)):*
> ```Shell
> 
> if [[ -n "${UD_SA_KEY_FILE}" ]]; then
>   ud::bazel_run //google/cloud/storage/tests:universe_domain_integration_test
> ```

As this is a test using gtest infrastructure, I would look at adding a `ud::bazel_test`, if necessary, or just using `bazel test`

**Comment by ddelgrosso1** [comment]
Realized I probably need to remove cmake from this due to #13495

**Review by ddelgrosso1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14728)* status: 0 of 7 files reviewed, 1 unresolved discussion (waiting on @scotthart)

___
*[`ci/cloudbuild/builds/universe-domain-storage.sh` line 29 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14728#-O7_kQHH1Dc8cStlfUc-:-O7dvYSA0I7q2FYp70Mh:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/13d08a440771d288e8f24db49f1c56ac34c06f8b/ci/cloudbuild/builds/universe-domain-storage.sh#L29)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

As this is a test using gtest infrastructure, I would look at adding a `ud::bazel_test`, if necessary, or just using `bazel test`
</blockquote></details>

Done.

File: google/cloud/storage/tests/BUILD.bazel [hunk 1]
```diff
@@ -54,3 +54,31 @@ VARIATIONS = {
         "@com_google_googletest//:gtest_main",
     ],
 ) for test in storage_client_integration_tests for v_label, v_env in VARIATIONS.items()]
+
+[cc_test(
+    name = test.replace("/", "_").replace(".cc", ""),
```
Comment (bajajneha27) [inline google/cloud/storage/tests/BUILD.bazel:59]: QQ: can you explain why we're doing this?

File: google/cloud/storage/tests/universe_domain_integration_test.cc [hunk 2]
```diff
@@ -0,0 +1,98 @@
+// Copyright 2022 Google LLC
```
Comment (bajajneha27) [inline google/cloud/storage/tests/universe_domain_integration_test.cc:1]: nit: Copyright year to be 2024

File: google/cloud/storage/tests/universe_domain_integration_test.cc [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  std::string const& bucket_name() const { return bucket_name_; }
+  std::string const& object_name() const { return object_name_; }
+
+ private:
+  std::string bucket_name_;
+  std::string object_name_;
+};
+
+auto TestOptions() {
+  auto sa_key_file = GetEnv("UD_SA_KEY_FILE").value_or("");
```
Comment (bajajneha27) [inline google/cloud/storage/tests/universe_domain_integration_test.cc:53]: How are we setting the value of these environment variables?

File: google/cloud/storage/tests/universe_domain_integration_test.cc (hunk 2 above)
Comment (ddelgrosso1) [reply google/cloud/storage/tests/universe_domain_integration_test.cc:1]: Thanks. This is what I get for copying and pasting 😆

File: google/cloud/storage/tests/BUILD.bazel (hunk 1 above)
Comment (ddelgrosso1) [reply google/cloud/storage/tests/BUILD.bazel:59]: My understanding is that this is creating the binary name by replacing the extension / slashes with the appropriate replacements. Looks like this is pretty consistent across our bazel files.

File: google/cloud/storage/tests/universe_domain_integration_test.cc (hunk 3 above)
Comment (ddelgrosso1) [reply google/cloud/storage/tests/universe_domain_integration_test.cc:53]: Actually your comment made me realize I had to actually pass these to bazel test. Updated the code.

**Review by bajajneha27 — APPROVED** [review]

**Review by scotthart — APPROVED** [review]
> Reviewed 3 of 3 files at r4, 3 of 4 files at r5, 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14728)* status: all files reviewed, 4 unresolved discussions (waiting on @bajajneha27)

**Pull request merged**